invenio/utils/deprecation.py

```diff
@@ -87,17 +90,61 @@ class RemovedInInvenio22Warning(DeprecationWarning):
     """Mark feature that will be removed in Invenio version 2.2."""
 
 
+# Modified version of http://stackoverflow.com/a/22376126/1727265
+def warn_with_traceback(message, category, filename, lineno, file=sys.stderr,
+                        line=None):
+    """Log deprecation warnings to debug logger."""
+    formatted_stack = traceback.format_stack()
+    ignore_path_matches = ('flask_debugtoolbar',)  # Regexes, look below.
+
+    log_this = True  # Until challenged
+    if any((re.match('  File .*('+'|'.join(ignore_path_matches)+').*', line)
+            for line in formatted_stack)):
+        log_this = False
+
+    # Format
+    if log_this:
+        message.args += 'Full backtrace in the logfile if DEBUG is enabled',
+    formatted_warning = warnings.formatwarning(message, category, filename,
+                                               lineno, line)
+
+    # Debug
+    # Yes, we want the augmented `message.args` so that people can / for it.
+    if log_this:
+        try:
+            current_app.logger.debug(formatted_warning)
+            current_app.logger.debug(''.join(formatted_stack))
+        except RuntimeError:
+            # App not yet initialized; see no evil, hear no evil.
+            message.args = message.args[:-1]
+            formatted_warning = warnings.formatwarning(message, category, filename,
+                                                       lineno, line)
+
+    # Warn
+    file.write(formatted_warning)
+
+
 # Improved version of http://code.activestate.com/recipes/391367-deprecated/
 def deprecated(message, category):
+    """Decorator which can be used to mark functions as deprecated.
+
+    ..note::
+        This function surrounds the call it makes to `warnings.warn` with
+        an on-the-fly patching of `warnings.showwarning`. This is so that
+        we can log a full stacktrace (without littering the terminal),
+        since it may be impossible to figure out what invoked the decorator
+        from the single stacklevel `warn` prints.
+
+    :param message: text to include in the warning
+    :param category: warning category
+    """
     def wrap(func=None):
-        """Decorator which can be used to mark functions as deprecated.
-
-        :param message: text to include in the warning
-        :param category: warning category
-        """
         @wraps(func)
         def new_func(*args, **kwargs):
+            original_showwarning = warnings.showwarning
```

Comment:
After a short chat before Jan left, we are leaning towards this solution:

* Do not manipulate the message logged to the console.
* Introduce a variable for enabling logging stacktraces of deprecationwarnings to the logfiles, for those interested in tackling them.

The problem of reproducibility remains, but we can live with it, since these stacktraces can make the logfile blow up in size tremendously.